### ISSUE TITLE
False positive for PossiblyNullArrayOffset

### ISSUE THREAD

**Issue by ErikBooijCB** [vimeo/psalm#515]
```php
if (isset($this->someProperty[$variable][$otherVariable])) {
    return $object->withProperty((string)$this->someProperty[$variable][$otherVariable]);
}
```

In a psalm run, this yields a `PossiblyNullArrayOffset` on the second line.

**Comment by muglug** [comment]
Reproducible with:

```php
$arr = [[1, 2, 3], null, [1, 2, 3], null];

$b = 2;
if (isset($arr[$b][0])) {
  echo $arr[$b][0];
}
```

The fix is to allow array variables with variable offsets into the list of known types, and to invalidate when any of the variable offsets is reassigned, or falls out of scope.

**Comment by muglug** [comment]
Thanks! Fixing this means Psalm can find issues like https://github.com/vimeo/psalm/commit/f56edd3c046012434985d5f3373b97847d3e233b now.
